src/pull-retry.ts

```diff
@@ -0,0 +1,94 @@
+/*!
+ * Copyright 2019 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {StatusObject, status} from 'grpc';
+
+export interface PullRetryOptions {
+  maxTimeout: number;
+}
+
+/*!
+ * retryable status codes
+ */
+const RETRY_CODES: status[] = [
+  status.OK,
+  status.CANCELLED,
+  status.UNKNOWN,
+  status.DEADLINE_EXCEEDED,
+  status.RESOURCE_EXHAUSTED,
+  status.ABORTED,
+  status.INTERNAL,
+  status.UNAVAILABLE,
+  status.DATA_LOSS,
+];
+
+/**
+ * Used to track pull requests and determine if additional requests should be
+ * made, etc.
+ *
+ * @class
+ * @private
+ */
+export class PullRetry {
+  private failures = 0;
+  private lastSuccess = Date.now();
+  private maxTimeout: number;
+  constructor(options = {} as PullRetryOptions) {
+    this.maxTimeout = options.maxTimeout;
+  }
+  /**
+   * Generates a timeout that can be used for applying a backoff based on the
+   * current number of failed requests.
+   *
+   * @private
+   * @returns {number}
+   */
+  createTimeout(): number {
+    if (this.failures === 0) {
+      return 0;
+    }
+    return Math.pow(2, this.failures) * 1000 + Math.floor(Math.random() * 1000);
+  }
+  /**
+   * Determines if a request status should be retried.
+   *
+   * Deadlines behave kind of unexpectedly on streams, rather than using it as
+   * an indicator of when to give up trying to connect, it actually dictates
+   * how long the stream should stay open. Because of this, it is virtually
+   * impossible to determine whether or not a deadline error is the result of
+   * the server closing the stream or if we timed out waiting for a connection.
+   * Typically we might use something like grpc's waitForReady, but there is an
+   * open issue that prevents this from working with channel pooling.
+   *
+   * @private
+   * @param {object} status The request status.
+   * @returns {boolean}
+   */
+  retry(s: StatusObject): boolean {
+    if (s.code === status.OK || s.code === status.DEADLINE_EXCEEDED) {
```

Comment:
there's discussion around auto-generated APIs, regarding making it so `DEADLINE_EXCEEDED` is no longer retried; the reasoning being it might not be idempotent:

1. are all the pubsub commands safe to retry on DEADLINE_EXCEEDED. (as an example, perhaps we're creating a new queue or something and we'd create two?).
2. should we add a comment for future humans, indicating that we should not remove this status check? even if we stop retrying `DEADLINE_EXCEEDED` in other APIs.

Reply:
I actually wrote a little blurb specifically about why we basically ignore deadline errors in the JSDoc description for this method. Deadline behavior changes a bit when applied to a stream, instead of being a timeout it actually dictates how long the stream should stay open for.

Reply:
:+1: missed it in the jsdoc.